tests/bluetooth/bsim/host/adv/encrypted/src/central.c

```diff
@@ -0,0 +1,90 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "common.h"
+
+#define EXPECTED_DATA_SIZE 6
+static uint8_t expected_data[] = {0x5A, 0x65, 0x70, 0x68, 0x79, 0x72};
+BUILD_ASSERT(sizeof(expected_data) == EXPECTED_DATA_SIZE);
+
+static uint8_t session_key[] = {0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7,
+				0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF};
+BUILD_ASSERT(sizeof(session_key) == BT_EAD_KEY_SIZE);
+static uint8_t iv[] = {0xFB, 0x56, 0xE1, 0xDA, 0xDC, 0x7E, 0xAD, 0xF5};
+BUILD_ASSERT(sizeof(iv) == BT_EAD_IV_SIZE);
+
+static bt_addr_le_t prev_addr;
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	int err;
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+
+	LOG_DBG("Device found: %s (RSSI %d)", addr_str, rssi);
+
+	if (!bt_addr_le_eq(addr, &prev_addr)) {
+		uint8_t *encrypted_data =
+			&ad->data[2]; /* the ad data are after the length and the type */
+		uint8_t encrypted_data_size = ad->len - 2; /* remove the type size */
+
+		uint8_t data_size =
+			((encrypted_data_size - BT_EAD_RANDOMIZER_SIZE) - BT_EAD_AAD_SIZE) -
+			BT_EAD_MIC_SIZE;
+		uint8_t *data = (uint8_t *)malloc(sizeof(uint8_t) * data_size);
```

Comment:
Should we use `k_malloc` and `k_free` instead?

Reply:
Not sure what the technical difference is, but it is suggested by the documentation: https://docs.zephyrproject.org/latest/develop/languages/c/index.html#dynamic-memory-management

Reply:
Just as info: in the bsim target, by default malloc will use the underlaying linux malloc(). Meaning, you are allocating memory in the linux process heap. While k_malloc uses the Zephyr heap (which is actually static memory in the linux process).
If you want something that would run in the embedded target and/or emulate the embedded behaviour you want to use k_malloc/k_free. If you just want a buffer for bsim specific test purposes you can use whatever you prefer.